src/proxy_protocol/mod.rs

```diff
@@ -0,0 +1,274 @@
+
+use crate::accept::Accept;
+use std::fmt;
+use tokio::io::{AsyncRead, AsyncWrite};
+
+pub mod future;
+use self::future::ProxyProtocolAcceptorFuture;
+
+pub trait Peekable {
+    fn poll_peek(
+        &mut self,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<std::io::Result<usize>>;
+}
+
+/// Note: Currently only supports the V2 PROXY header format.
+/// See proxy header spec: <https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt>
+pub(crate) async fn read_proxy_header<I>(mut stream: I) -> Result<(I, Option<IpAddr>), io::Error>
+where
+    I: Peekable + AsyncRead + Unpin,
+{
+    // Maximum expected size for PROXY header is
+    // unlikely to be larger than 512 bytes unless additional fields are added.
+    const BUFFER_SIZE: usize = 512;
+    // Mutable buffer for storing the bytes received from the stream.
+    let mut array_buffer = [0; BUFFER_SIZE];
+    let mut buffer = ReadBuf::new(&mut array_buffer);
+
+    // peek the stream until find a complete header
+    let header = loop {
+        // Peek the stream data into the buffer.
+        // Each loop re-reads the stream from beginning,
+        // but further in if more stream has arrived.
+        poll_fn(|cx| stream.poll_peek(cx, &mut buffer)).await?;
+
+        let header = HeaderResult::parse(buffer.filled());
+
+        if header.is_complete() {
+            break header;
+        } else if buffer.remaining() == 0 {
+            // Buffer limit reached without finding a complete header. Consider increasing the buffer size.
+            return Ok((stream, None));
+        }
+
+        buffer.clear();
+    };
+
+    match header {
+        HeaderResult::V1(Ok(_header)) => {
+            // V1 PROXY header detected, which is not supported
+            Ok((stream, None))
+        }
+        HeaderResult::V2(Ok(header)) => {
+            let client_address = match header.addresses {
+                v2::Addresses::IPv4(ip) => IpAddr::V4(ip.source_address),
+                v2::Addresses::IPv6(ip) => IpAddr::V6(ip.source_address),
+                v2::Addresses::Unix(_unix) => {
+                    // Unix socket addresses are not supported
+                    return Ok((stream, None));
+                }
+                v2::Addresses::Unspecified => {
+                    // V2 PROXY header addresses "Unspecified"
+                    return Ok((stream, None));
+                }
+            };
+
+            // Remove header length in bytes from the stream
+            let header_len = header.len();
+            stream
+                .read_exact(&mut buffer.filled_mut()[..header_len])
+                .await?;
+
+            Ok((stream, Some(client_address)))
+        }
+        HeaderResult::V1(Err(_error)) => {
+            // No valid V1 PROXY header received
+            Ok((stream, None))
+        }
+        HeaderResult::V2(Err(_error)) => {
+            // No valid V2 PROXY header received
+            Ok((stream, None))
```

Comment:
Why is this OK?

Reply:
was to leave the stream unharmed because it likely means the proxy header isn't there, is what i get when i give it a vanilla stream without the header